BUG #19860 Corrected use of mixed indexes with .at

mariuspot opened the pull request:
`.at` incorrectly disallowed the use of integer indexes when a mixed index was used
```
s = Series([1,2,3,4,5],index=['a','b','c',1,2])
s.at['a'] # returns 1
s.at[1]   # returns 4
s.at[4]   # raises KeyError, doesn't do fallback indexing
```

Made sure fallback indexing doesn't happen on mixed indexes

- [X] closes #19860
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

mariuspot committed:
BUG #19860 Corrected use of mixed indexes with .at

.at incorrectly disallowed the use of integer indexes when a mixed index
was used
disabled fallback indexing when a mixed index is used

jreback commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -666,6 +666,20 @@ def test_index_type_coercion(self):
                     idxr(s2)['0'] = 0
                     assert s2.index.is_object()
 
+    @pytest.mark.parametrize("index,val", [
```
can you add a comment with the issue number

jreback replied:
these tests should be in pandas/tests/indexes/test_base (look for other contains tests)

mariuspot replied:
`test_index_not_contains` and `test_index_contains` are in the same file a few lines above, should I move these up or create pandas/tests/indexes/test_base and move them there?

jreback commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -710,6 +724,27 @@ def test_float_index_at_iat(self):
         for i in range(len(s)):
             assert s.iat[i] == i + 1
 
+    def test_mixed_index_at_iat(self):
```
move these to pandas/tests/indexing/test_scalar.py (see where other ones are)

jreback replied:
add a gh comment with the issue number

jreback replied:
can you add the test from the OP (was on dataframe)

jreback commented on pandas/tests/indexing/test_indexing.py:
```diff
@@ -710,6 +724,27 @@ def test_float_index_at_iat(self):
+        s = Series([1, 2, 3, 4, 5], index=['a', 'b', 'c', 1, 2])
+        for el, item in s.iteritems():
+            assert s.at[el] == item
+        for i in range(len(s)):
```
can you check that we have similar tests for .loc (I think so), if not can you add a loc version (inside same test is ok). maybe even just add the .loc tests and move from elsewhere.

you can parameterize these over .at / .loc

jreback reviewed (changes requested):
looks pretty good. can you add a whatsnew note in bugfixes indexing for 0.24

jreback commented:
also there *might* be other .at issues which this closes. pls have a look if you can.

cc @toobaz

mariuspot committed:
Fixed issues from review

Added whatsnew to 0.24
Added issue to tests comments
Moved some test
Added a test from the issue

pep8speaks commented:
Hello @mariuspot! Thanks for updating the PR.







 - In the file [`pandas/tests/indexing/test_scalar.py`](https://github.com/pandas-dev/pandas/blob/46025b0cdfcd948a3087dcda0202a9602cb540a4/pandas/tests/indexing/test_scalar.py), following are the PEP8 issues :

> [Line 188:5](https://github.com/pandas-dev/pandas/blob/46025b0cdfcd948a3087dcda0202a9602cb540a4/pandas/tests/indexing/test_scalar.py#L188): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 191:27](https://github.com/pandas-dev/pandas/blob/46025b0cdfcd948a3087dcda0202a9602cb540a4/pandas/tests/indexing/test_scalar.py#L191): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 205:1](https://github.com/pandas-dev/pandas/blob/46025b0cdfcd948a3087dcda0202a9602cb540a4/pandas/tests/indexing/test_scalar.py#L205): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

mariuspot committed:
Fixed PEP8 issues :(

jreback commented:
looks good. can you rebase

jreback commented:
looks like you picked up some other commits, do this
```git merge upstream/master`` and push again

mariuspot commented:
`git merge upstream/master` didn't do anything
Not sure what I did now :/
Hope it's ok now

mariuspot committed:
Merge remote-tracking branch 'upstream/master' into issue-19860

mariuspot commented:
Ok, I figured out what went wrong with the merge. Should be ok now

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -659,7 +659,11 @@ Indexing
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
 - Bug where indexing with a Numpy array containing negative values would mutate the indexer (:issue:`21867`)
+<<<<<<< HEAD
```
hve a rebase issue here

jreback reviewed (changes requested):
lgtm. looks like some extraneous rebase issue. ping on green.

mariuspot committed:
Fixed rebase issue

jreback commented:
lgtm ping on green.

mariuspot commented:
Looks like travis failed with a segfault on the linker

mariuspot commented:
@jreback The travis build had a segfault with the linker while setting up the environment. What is the process for that. Is there a way for me to rerun the build or should I make a change and do a new commit?

mariuspot committed:
Merge branch 'master' of github.com:pandas-dev/pandas into issue-19860

jreback commented:
thanks!
